File: python/average_levels_of_binary_tree.py
# Definition for a binary tree node.
from typing import List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        result = [0]
        queue = deque()
        queue.append(root)
        queue.append(None)
        count = 0
        while queue:
            current = queue.popleft()
            if current is None and len(queue) == 0: break
            if current is None:
                queue.append(None)
                result[-1] /= count
                result.append(0)
                count = 0
                continue
            count += 1
            if current.left is not None: queue.append(current.left)
            if current.right is not None: queue.append(current.right)
            result[-1] += current.val
        result[-1] /= count
        return result

File: src/AverageLevelsOfBinaryTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class AverageLevelsOfBinaryTree {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        queue.add(null);
        int count = 0;
        List<Double> result = new ArrayList<>();
        result.add(0.0);
        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            if (current == null && queue.isEmpty()) break;
            if (current == null) {
                queue.add(null);
                result.set(result.size() - 1, result.get(result.size() - 1) / count);
                result.add(0.0);
                count = 0;
                continue;
            }
            count++;
            if (current.left != null) queue.add(current.left);
            if (current.right != null) queue.add(current.right);
            result.set(result.size() - 1, result.get(result.size() - 1) + current.val);
        }
        result.set(result.size() - 1, result.get(result.size() - 1) / count);
        return result;
    }
}
